File: clp/haskell/exercicios.hs
-- Resposta da lista de exercícios (versão Haskell)

-- 1. Implemente uma função/regra que verifica
--    se um elemento pertence a uma lista.

pertence x [] = False
pertence x (a:b) = if a == x then True else pertence x b

--essa questão equivale à função elem do haskell 

pertence2 x lista = elem x lista


-- 2. Implemente uma função/regra que inverte
--    os elementos de uma lista.

inverte [] = []
inverte (a:b) = inverte b ++ [a]

--essa questão equivale à função reverse do haskell 

inverte2 lista = reverse lista


-- 3. Implemente uma função/regra que informe
--    o n-ésimo membro de uma lista.

n_esimo 0 (a:_) = a 
n_esimo i (a:b) = n_esimo (i-1) b

--essa questão equivale ao operador !! do haskell 

n_esimo2 x lista = lista !! x

-- 4. Implemente uma função/regra que remova
--    todas as ocorrências de um elemento em uma lista.

retiraTodos a lst = [x | x <- lst, x /= a]

retiraTodos2 a lst = filter (/=a) lst


-- 5. Implemente uma função/regra que recebe 
--    uma lista e a divide em duas pela metade.

divide lista = (take metade lista, drop metade lista)
   where
      metade = div (length lista) 2
 

-- 6. Implemente uma função/regra que verifica se uma String
--    é ou não um palíndromo. Um palíndromo é uma String que
--    se invertida se mantém exatamente igual à original.

palindromo lista = lista == (inverte lista)   


-- 7. Implemente uma função/regra primo(N) que dado um número
--    inteiro N, determine se ele é ou não um número primo.
--    round arredonda um float e sqrt retorna a raiz de um número.
--OBS.: a maioria dos autores falam que 1 não é primo

primo :: Integer -> Bool
primo 0 = False
primo 1 = False
primo 2 = True
primo x = primo_aux x (truncate (sqrt (fromIntegral x)))

primo_aux :: Integer -> Integer -> Bool 
primo_aux _ 1 = True
primo_aux x d = if mod x d == 0 then False else primo_aux x (d-1)   


-- 8. Implemente uma função/regra lista_primos(N, Lista) que
--    aceite um número inteiro e que determine a lista de todos
--    os números primos iguais ou inferiores a esse número.


lista_primos n
  | n < 2     = []  
  | primo(n)  = lista_primos (n-1) ++ [n]
  | otherwise = lista_primos (n-1) 


-- 9. Escreva uma função/regra chamada conceito que receba
--    uma nota como argumento e retorne o conceito
--    correspondente, conforme as regras:  nota >= 9.0,
--    conceito A; nota >= 7.5 e < 9.0, conceito B; nota >= 6.0
--    e < 7.5, conceito C; nota >= 4.0 e < 6.0, conceito D;
--    nota < 4.0, conceito E.

conceito nota 
  | nota >= 9.0 = 'A'
  | nota >= 7.5 && nota < 9.0 = 'B'
  | nota >= 6.0 && nota < 7.5 = 'C'
  | nota >= 4.0 && nota < 6.0 = 'D'
  | nota < 4.0 = 'E'


-- 10. Implemente o algoritmo de ordenação mergesort.

mergesort []  = []
mergesort [x] = [x]
mergesort lst = conquistar (mergesort me) (mergesort md)
  where (me,md) = divide lst

conquistar [] lst = lst
conquistar lst [] = lst
conquistar (p1:r1) (p2:r2)
  | p1 < p2   = p1 : conquistar r1 (p2:r2)
  | otherwise = p2 : conquistar (p1:r1) r2 

File: clp/prolog/exercicios.pl
% Resposta da lista de exercicios (versao Prolog)

% 1. Implemente uma funcao/regra que verifica
%    se um elemento pertence a uma lista.
pertence(X,[X|_]) :- !.
pertence(X,[_|Y]) :- pertence(X,Y).


% 2. Implemente uma funcao/regra que inverte
%    os elementos de uma lista.

inverte(L,Li) :- inverte(L,[],Li).
inverte([],L,L).
inverte([X|Y],Li,L) :- inverte(Y,[X|Li],L).

% Outra maneira de fazer usando a regra append
% append (fornecida pelo Prolog) concatena duas listas

inverte2([],[]).
inverte2([X|Y],Z) :- inverte2(Y,Yi), append(Yi,[X],Z).


% 3. Implemente uma funcao/regra que informe
%    o n-esimo membro de uma lista.

n_esimo(0,[X|_],X).
n_esimo(I,[_|Y],Z) :- N is I-1, n_esimo(N,Y,Z).


% 4. Implemente uma funcao/regra que remova
%    todas as ocorrencias de um elemento em uma lista.

removeTodos(_,[],[]).
removeTodos(X,[X|Y],Z) :- removeTodos(X,Y,Z).
removeTodos(X,[N|Y],[N|Z]) :- removeTodos(X,Y,Z).


% 5. Implemente uma funcao/regra que recebe 
%    uma lista e a divide em duas pela metade.

divide(L,E,D) :- length(L,T), M is T div 2, divide(L,M,E,D), !.
divide(L,0,[],L) :- !.
divide([X|Y],P,[X|E],D) :- Z is P-1, divide(Y,Z,E,D).
 

% 6. Implemente uma funcao/regra que verifica se uma String
%    eh ou nao um palindromo. Um palindromo eh uma String que
%    se invertida se mantem exatamente igual a original.
%    string_chars converte a string para lista de chars.

palindromo(X) :- string_chars(X,L), inverte(L,Y), L=Y. 


% 7. Implemente uma funcao/regra primo(N) que dado um numero
%    inteiro N, determine se ele eh ou nao um numero primo.
%    round arredonda um float e sqrt retonra a raiz de um numero.
%  OBSERVACOES IMPORTANTES: 
%  A maioria dos autores falam que 1 nao eh primo.
%  Nao eh possivel usar o resultado de uma operacao ou regra 
%  como entrada de uma regra. Por exemplo, nao eh permitido escrever 
%  fatorial (X-1). Mas algumas regras fornecidas pelo Prolog
%  atuam como funcoes que tem retorno quando usadas junto com 
%  o operador is. Por exemplo, sqrt(X,Y) eh uma regra em que Y 
%  eh a raiz quadrada de X e round(X,Y) eh uma regra em que Y eh
%  o X arredondado. Usar R is round(sqrt(X)) eh o mesmo que 
%  sqrt(X,Z),round(Z,R). Regras que criamos nao podem fazer isso.   


primo(X) :- X >= 2, R is round(sqrt(X)), primo(X,R).
primo(X,1) :- !.
primo(X,D) :- R is X mod D, R \= 0, Y is D-1, primo(X,Y).


% 8. Implemente uma funcao/regra lista_primos(N, Lista) que
%    aceite um numero inteiro e que determine a lista de todos
%    os numeros primos iguais ou inferiores a esse numero.

lista_primos(0,[]) :- !.
lista_primos(1,[]) :- !. 
lista_primos(2,[2]) :- !.
lista_primos(N,Lista) :- primo(N),
   Z is N-1, lista_primos(Z,Y), append(Y,[N], Lista), !.
lista_primos(N,Lista) :- Z is N-1, lista_primos(Z,Lista).


% 9. Escreva uma funcao/regra chamada conceito que receba
%    uma nota como argumento e retorne o conceito
%    correspondente, conforme as regras:  nota >= 9.0,
%    conceito A; nota >= 7.5 e < 9.0, conceito B; nota >= 6.0
%    e < 7.5, conceito C; nota >= 4.0 e < 6.0, conceito D;
%    nota < 4.0, conceito E.

conceito(Nota,'A') :- Nota >= 9.0.
conceito(Nota,'B') :- Nota >= 7.5, Nota < 9.0.
conceito(Nota,'C') :- Nota >= 6.0, Nota < 7.5.
conceito(Nota,'D') :- Nota >= 4.0, Nota < 6.0.
conceito(Nota,'E') :- Nota < 4.0.

% 10. Implemente o algoritmo de ordenacao mergesort.

mergesort([],[]) :- !.
mergesort([X],[X]) :- !.
mergesort(L,Lo) :- divide(L,E,D), 
  mergesort(E,Eo), mergesort(D,Do), conquista(Eo,Do,Lo).

conquista([],L,L) :- !. 
conquista(L,[],L) :- !. 
conquista([X1|R1],[X2|R2],[X1|L]) :- X1 < X2, conquista(R1,[X2|R2],L), !. 
conquista([X1|R1],[X2|R2],[X2|L]) :- X1 >= X2, conquista([X1|R1],R2,L), !. 
